lib/galaxy/datatypes/genetics.py

```diff
@@ -827,6 +829,637 @@ def __init__(self, **kwd):
+    def tokenizer(line, sep=None):
+        """
+        General purpose string tokenizer
+        """
+        if sep is None:
+            return line.splitlines()[0].split(sep)
+
+        return line.splitlines()[0].split(sep)
+
+    def eof_function(self):
+        """
+        Overridable end-of-file function
+        """
+        return self.eof_res
+
+    def __per_line_op(self, line):
+        """
+        Private per-line operation and line counter
+        """
+        self.lcount += 1
+        if self.lcount > self.max_lines:
+            return -1
+
+        return self.line_op(line)
+
+    def line_op(self, line):
+        """
+        Overridable per line operation
+        """
+        return None
+
+    def header_check(self, fio):
+        """
+        Overrideable post-binary file check function
+        """
+        return True
+
+    def sniffer(self, filename):
+        with open(filename, "r") as fio:
+
+            if LinkageStudies.__is_binary_file(fio):
+                return False
+
+            if not self.header_check(fio):
+                return False
+
+            for line in fio:
+                line_res = self.__per_line_op(line)
+
+                if line_res == -1:
+                    # run eof_function
+                    break
+
+                if line_res is not None:
+                    return line_res
+
+            return self.eof_function()
+        return False
+
+
+class PreMakePed(LinkageStudies):
+    """
+    Common linkage pedin format
+    Extended linkage pedigree file containing:
+         fam, indiv, fath, moth, gender, affectation, [genotypes]
+    """
+    file_ext = "linkage_pedin"
+
+    def __init__(self, **kwd):
+        LinkageStudies.__init__(self, **kwd)
+        self.num_colns = None
+        self.max_lines = 10
+        self.fam_id = -1
+        self.cols_found = -1
+
+    def line_op(self, line):
+        tokens = LinkageStudies.tokenizer(line)
+
+        if line.startswith("   "):
+            return False
+
+        if self.num_colns is not None:
+            if len(tokens) != self.num_colns:
+                return False
+
+        else:
+            if self.cols_found == -1:
+                self.cols_found = len(tokens)
+                # absolute minimum for pedinfo + alleles
+                if self.cols_found < 7:
+                    return False
+
+            elif self.cols_found != len(tokens):
+                return False
+
+        try:
+            if set([int(val) >= 0 for val in tokens]) != {True}:
+                return False
+
+            if self.fam_id == -1:
+                self.fam_id = int(tokens[0])
+            elif self.fam_id != int(tokens[0]):
+                return False
+
+        except ValueError:
+            return False
+
+        return None
+
+    def sniff(self, filename):
+        """
+        >>> cname = "PreMakePed"
+        >>> from galaxy.datatypes.sniff import get_test_fname
+        >>> extn_true = eval(cname)().file_ext
```

Comment:
Don't over engineer things, you know all the extensions and they should stay stable.
xref: https://stackoverflow.com/questions/1832940/why-is-using-eval-a-bad-practice

Reply:
Sorry, this was leftover back when I was trying to implement inheritable doc-tests. Will cleanup asap